### DISCUSSION TITLE
usocket implementation to detect when a client disconnects

### DISCUSSION THREAD

**Discussion by Hruthik0x** [micropython/micropython#13423] [Libraries & Drivers]
Someone please show me an implementation of micropython code to setup a socket server that can handle multiple clients simultaneously and can detect when a client connects, disconnects.
There should be a mechanism, to detect which client disconnected.

**Comment by sophiedegran** [comment]
Hi
Did you look at github.com/peterhinch/micropython-async/blob/master/v3/as_drivers/client_server ?
For me the timeout is good.. you can decrease it..
Hope that help

**Comment by Hruthik0x** [comment]
But for this, I have to keep emitting text for every timeout seconds right ?
Isn't that polling ?

**Reply by sophiedegran** [reply]
I don't know what your expect to do with opening socket.. In this example userver wait for an exchange by uclient and if the timeout  exceeded userver close the socket connexion..
Perharps it's your client that have been to run userver?

**Comment by Hruthik0x** [comment]
In my case, I just want the server to know when a user connects and
disconnects from the server.
Socketio, Flask_socketio have something like
@socket.on("connect") and @socket.on ("disconnect") to detect when a client
connects and disconnects, they also provide the sid, by using request.sid
(unique to each client, could be used to identify each client and send
message specific to a client)
I am looking for something like that in micropython.

**Reply by sophiedegran** [reply]
Yes ok it's like websocket..
You will find them under microdot (it's framework web) under micropython as a 'small' flask 
but there's not directly this function to know if connect/ disconnect...
Have a look to microdot it's an amazing web server !
You can also type google 'websocket under micropython'☺

**Comment by Hruthik0x** [comment]
Thank you.
I'm planning to contribute to micropython, please tell me where to start (I
am new to micropython, so I want to know how to write code to contribute,
like build a new library etc.)

**Reply by sophiedegran** [reply]
Me I'm only try to give a good orientation for questions.. when I can... Rarely I will push code ... Read most answer from people on the right list and you will have the answer.. World of micropython is amazing ☺
